daft/dataframe/dataframe.py

```diff
@@ -3750,3 +3694,24 @@ def map_groups(self, udf: Expression) -> "DataFrame":
 
         """
         return self.df._map_groups(udf, group_by=self.group_by)
+
+
+T = TypeVar("T")  # Result type returned by `WriteSink.write()` and accepted by `WriteSink.finish()`.
+
+
+class WriteSink(ABC, Generic[T]):
+    """Interface for writing data to a sink that is not built-in."""
+
+    def start(self) -> None:
+        """Optional callback for when a write starts."""
+        pass
+
+    @abstractmethod
+    def write(self, dataframe: DataFrame, **kwargs) -> T:
```

Comment:
notes from discussion:we need to support use cases where some customized post processing needed to be done after the data file is written (in our case, tagging the s3 files created). I think we can do that within the `finish()` interface as part of our earlier release and defer that feature requests as later optimization.

Some thoughts on how that might be achieved [food for thought, not high priority]: add another hook like `post_batch_write()` which will be invoked when a batch of data has finished writting, so that customer can implement this hook and inject custom logic as needed.